core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -2348,7 +2348,20 @@ class KafkaController(val config: KafkaConfig,
           controllerContext.partitionLeadershipInfo(tp) match {
             case Some(leaderIsrAndControllerEpoch) =>
               val currentLeaderAndIsr = leaderIsrAndControllerEpoch.leaderAndIsr
-              if (newLeaderAndIsr.leaderRecoveryState == LeaderRecoveryState.RECOVERING && newLeaderAndIsr.isr.length > 1) {
+              if (newLeaderAndIsr.leaderEpoch < currentLeaderAndIsr.leaderEpoch) {
+                partitionResponses(tp) = Left(Errors.FENCED_LEADER_EPOCH)
+                None
+              } else if (newLeaderAndIsr.leaderEpoch > currentLeaderAndIsr.leaderEpoch) {
+                partitionResponses(tp) = Left(Errors.UNKNOWN_LEADER_EPOCH)
```

Comment:
Good questions. Another option I was considering is to call this `INVALID_REQUEST` since the controller is expected to have the latest leader epoch. I think `UNKNOWN_LEADER_EPOCH` is typically used in cases where we expect the failure to be transient.

Reply:
So what I ended up doing is using FENCED_LEADER_EPOCH for any case where the leader epoch does not match the current value. This is consistent with `ReplicationControlManager` and it is already handled on the broker side in `Partition` (unlike `UKNOWN_LEADER_EPOCH`).